packages/wallet/src/services/WalletUtil.ts

```diff
@@ -23,11 +23,79 @@ export interface WalletOutputValidatorContext {
 export type WalletUtilContext = WalletOutputValidatorContext & InputResolverContext;
 
 export const createInputResolver = ({ utxo }: InputResolverContext): Cardano.InputResolver => ({
-  async resolveInput(input: Cardano.TxIn) {
+  async resolveInput(input: Cardano.TxIn, options?: Cardano.ResolveOptions) {
     const utxoAvailable = await firstValueFrom(utxo.available$);
     const availableUtxo = utxoAvailable?.find(([txIn]) => txInEquals(txIn, input));
-    if (!availableUtxo) return null;
-    return availableUtxo[1];
+
+    if (availableUtxo) return availableUtxo[1];
+
+    if (options?.hints) {
+      const tx = options?.hints.find((hint) => hint.id === input.txId);
+
+      if (tx && tx.body.outputs.length > input.index) {
+        return tx.body.outputs[input.index];
+      }
+    }
+    return null;
+  }
+});
+
+/**
+ * Creates an input resolver that fetch transaction inputs from the backend.
+ *
+ * This function tries to fetch the transaction from the backend using a `ChainHistoryProvider`. It
+ * also caches fetched transactions to optimize subsequent input resolutions.
+ *
+ * @param provider The backend provider used to fetch transactions by their hashes if
+ * they are not found by the inputResolver.
+ * @returns An input resolver that can fetch unresolved inputs from the backend.
+ */
+export const createBackendInputResolver = (provider: ChainHistoryProvider): Cardano.InputResolver => {
+  const txCache = new Map<Cardano.TransactionId, Cardano.Tx>();
+
+  const fetchAndCacheTransaction = async (txId: Cardano.TransactionId): Promise<Cardano.Tx | null> => {
+    if (txCache.has(txId)) {
+      return txCache.get(txId)!;
+    }
+
+    const txs = await provider.transactionsByHashes({ ids: [txId] });
+    if (txs.length > 0) {
+      txCache.set(txId, txs[0]);
```

Comment:
Could this be caching transactions that will later be rolled back?
We could cache only if it's outside of the stability window.

Reply:
I also think it is harmless to cache transactions that re rolled back, because the point is to retrieve the information to properly display what the transaction is doing, if the user is referencing inputs from such transactions and there are more than one input from within the transaction we still save some fetching by caching it, because the input resolver only resolves one input at a time.

Does the input selector gets destroyed when the extension closes?, or whats the lifecycle of the input selector? if it is very long lived, I agree we need to limit the size of the cache just in case, altho very unlikely this cache will ever be too high since it only fetches foreign TXs and most inputs will be from the wallet most of the time. Anyways, let me know how strong you guys feel about this and I can add this extra change. We could use something like https://www.npmjs.com/package/lru-cache

Reply:
> Does the input selector gets destroyed when the extension closes?, or whats the lifecycle of the input selector?

It does when the service worker dies, so after like 10 or 15 minutes of inactivity.

> Anyways, let me know how strong you guys feel about this and I can add this extra change.

I don't think it's worth managing this cache as it's not likely that it would take up lots of memory